File: weyl-core/src/test/java/io/quartic/weyl/core/geojson/GeoJsonShould.java
package io.quartic.weyl.core.geojson;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import org.hamcrest.Matchers;
import org.junit.Assert;
import org.junit.Test;

import java.io.StringWriter;
import java.util.List;
import java.util.Optional;

public class GeoJsonShould {

    public static <T> List<T> list(T... stuff) {
        return ImmutableList.copyOf(stuff);
    }

    @Test
    public void deserializeBackToOriginal() throws Exception {

        FeatureCollection original = FeatureCollection.of(list(
                Feature.of(
                        Optional.empty(),
                        Point.of(list(102.0, 0.5)),
                        ImmutableMap.of()
                ),
                Feature.of(
                        Optional.empty(),
                        LineString.of(list(
                                list(102.0, 0.0),
                                list(103.0, 1.0),
                                list(104.0, 0.0),
                                list(105.0, 1.0)
                                )),
                        ImmutableMap.of()
                ),
                Feature.of(
                        Optional.empty(),
                        Polygon.of(list(list(
                                list(100.0, 0.0),
                                list(101.0, 0.0),
                                list(101.0, 1.0),
                                list(100.0, 1.0),
                                list(100.0, 0.0)
                        ))),
                        ImmutableMap.of()
                )
        ));

        ObjectMapper mapper = new ObjectMapper();
        mapper.registerModule(new Jdk8Module());
        mapper.enable(SerializationFeature.INDENT_OUTPUT);

        StringWriter sw = new StringWriter();
        mapper.writeValue(sw, original);

        String json = sw.toString();

        Assert.assertThat(mapper.readValue(json, FeatureCollection.class), Matchers.equalTo(original));
    }
}

File: common-core/src/test/kotlin/io/quartic/common/geojson/GeoJsonShould.kt
package io.quartic.common.geojson

import com.fasterxml.jackson.databind.SerializationFeature
import io.quartic.common.serdes.OBJECT_MAPPER
import org.hamcrest.Matchers.equalTo
import org.junit.Assert.assertThat
import org.junit.Test
import java.io.StringWriter

class GeoJsonShould {

    @Test
    fun deserializeBackToOriginal() {
        val original = FeatureCollection(listOf(
                Feature(geometry = Point(listOf(102.0, 0.5))),
                Feature(geometry = LineString(listOf(
                                listOf(102.0, 0.0),
                                listOf(103.0, 1.0),
                                listOf(104.0, 0.0),
                                listOf(105.0, 1.0)
                        ))),
                Feature(geometry = Polygon(listOf(listOf(
                                listOf(100.0, 0.0),
                                listOf(101.0, 0.0),
                                listOf(101.0, 1.0),
                                listOf(100.0, 1.0),
                                listOf(100.0, 0.0)
                        ))))
        ))

        OBJECT_MAPPER.enable(SerializationFeature.INDENT_OUTPUT)

        val sw = StringWriter()
        OBJECT_MAPPER.writeValue(sw, original)

        val json = sw.toString()

        assertThat(OBJECT_MAPPER.readValue(json, FeatureCollection::class.java), equalTo(original))
    }
}
